Draft: Implement Polyspace checker (#142)
melexis/warnings-plugin: README.rst

.. |--xunit report.xml| replace:: ``--xunit report.xml``
.. _`--xunit report.xml`: https://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#xunit-compatible-result-file

Parse for Polyspace Failures
----------------------------

The Polyspace checker requires as input a TSV file exported by Polyspace.
You can find instructions on exporting TSV files in the Polyspace documentation.

Exporting Polyspace Results
^^^^^^^^^^^^^^^^^^^^^^^^^^^

The following commands instruct Polyspace to export the results as a TSV file:

.. code-block:: bash

    polyspace-results-export -format csv -results-dir <resultsFolder> <export_options>
    # or
    polyspace-results-export -format csv -host <hostName> -run-id <runID> <export_options> <polyspace_access_options>

The csv format outputs tab-separated values (TSV).
Each result in the TSV format consists of tab-separated information in columns.
Starting from Polyspace version R2024a these columns are available:
ID, Family, Group, Color, New, Check, Information, Function, File, Status, Severity, Comment, Key, Line and Col.
In previous versions of Polyspace the Line and Col column are not available yet.

> This file is necessary when you enable Polyspace in the configuration file.

Reviewer: ```suggestion
This file is required when you enable Polyspace in the configuration file.
```

Configuration
^^^^^^^^^^^^^

> Polyspace checking can only be used with a configuration file,

Reviewer: ```suggestion
Polyspace checking can only be enabled with a configuration file,
```

.. |--xunit report.xml| replace:: ``--xunit report.xml``
.. _`--xunit report.xml`: https://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#xunit-compatible-result-file

Parse for Polyspace Failures
----------------------------

> To use the Polyspace checker, a TSV file must be exported from Polyspace.

Author: Resolved in 48647aa2d9fed78a2a90943414329e982a5305da

You can find instructions on exporting TSV files in the Polyspace documentation.

> Exporting the TSV File
> ^^^^^^^^^^^^^^^^^^^^^^

Reviewer 2: ```suggestion
Exporting Polyspace Results
^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

The following commands export a TSV file from Polyspace:

.. code-block:: bash

    polyspace-results-export -format csv -results-dir <resultsFolder> <export_options>
    # or

>     polyspace-results-export -format csv -host <hostName> -run-id <runID> <export_options> <polyspace_access_options>

Reviewer 2: Let's include the `-key-mode function-scope` if it proves to be useful

Author: Resolved in f9ad210e1a09e4bf230647604f141663d36d7f0b

Reviewer 2: It has proven to be useless for our use case.

Author: Reverted in a2724d06e77dac0ad4358f78d96cc1afa6404128

The csv format outputs tab-separated values (TSV).
Each result in the TSV format consists of tab-separated information in columns such as ID,

> Family, Group, Color, Check, and so on.

Reviewer 2: Would it be helpful to document all column titles here? And mention that, starting from R2024a, line and column information are also exported.

Author: I can say that starting from R2024a columns: ID, Family, Group, Color, New, Check, Information, Function, File, Status, Severity, Comment, Key, Line and Col are available

Author: Resolved in dd3b13e7d55e4336828d41761d9d4c67902b505c

Reviewer 2: You're not documenting the distinction between pre and post R2024a. This could be useful information.

Author: Is this sufficient? aee25d04f9ea9afcc3fa124a444fd3adb900d136

This file is necessary when you enable Polyspace in the configuration file.

Configuration
^^^^^^^^^^^^^

Polyspace checking can only be used with a configuration file,
and it cannot be used together with other checkers enabled.
In this case, only the Polyspace checker will run.

When you enable Polyspace checking in the configuration file,
the checks consist of a key that represents the "family" column of the TSV file.
For example, "run-time check" is the family of Code Prover and "defect" is the family of Bug Finder.
The value of that key is a list, which contains the name of the column to check as a key and
the value of that column to check together with ``min`` and ``max`` values.

Example Checks
^^^^^^^^^^^^^^

In case of Code Prover, you might want to check the ``color`` column on ``red`` or ``orange`` issues.
In case of Bug Finder, you might want to check the ``information`` column on ``impact: high``, ``impact: medium``, or even ``impact: low``.
Other issues, such as "Global variable", can also be handled.
You can specify any column and value you want to check in the configuration file.

Default Description Template
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> ``Polyspace: $check`` is the default value for the ``cq_description_template`` variable.

Author: Resolved in ca768433f10eabf26217de0eaee662ec8994d83c

                    "max": 0
                }
            ]
        },
        "polyspace": {
            "enabled": false,
            "cq_description_template": "$PRODUCT $family: $check",
            "exclude": [

>                 ".+dummy_function\(\)|dummy_file_name\.c"

Author: Apparently these backslashes are not necessary. The tests for testing the excluded has the same result. Fixed the red color in 98ca303179cc43f2f87b9acc2f984fa94a1e57ee

>                 ".+dummy_function\(\)|dummy_file_name\.c"

Author: So in order to have a `\` when JSON is handled by Python I need to use `\\`

>                 ".+dummy_function\(\)|dummy_file_name\.c"

Author: In this case the tab can be used but when there was a row in between, `|` can be used.

>                 ".+dummy_function\(\)|dummy_file_name\.c"

Author: Resolved in a3a571c60f8a00839faae620302f2c3ccc9a32cf
Added documentation in 76f337fd4d5e80aaad3c6014bb417834e334f6b1

        }
    }

Polyspace checker lets you define patterns (regular expressions) to exclude specific rows of the TSV file.
These patterns are compared against each tab-separated string that represents a row.
This way, you can combine cell values with tabs in the regex.
Remember, backslashes need escaping in JSON configurations but not in YAML for these patterns.

.. code-block:: json

    {
        "polyspace": {

>             "enabled": false,

Reviewer 2: ```suggestion
            "enabled": true,
```

If not configured, ``.gitlab-ci.yml`` will be used as a fallback path.

You can customize the description with ``"cq_description_template"``, see `configuration file to pass options`_.
Its value should be a template for Python's |string.Template|_. The template has access to all environment variables,
e.g. ``$HOME``, and other variables that depend on the checker type:

Polyspace
  Any field of a Polyspace defect can be included by using the corresponding

>   `column title <Exporting Polyspace Results>`_ in lowercase as the variable name.

Reviewer 2: I vaguely remember that [Implicit Hyperlink Targets for Section Titles](https://gist.github.com/jwodder/5a9264db8c3c81dad4c01e082515a9f1#id13) are not supported on GitHub. We should use an alternative method to reference a title internally.

Author: Fixed the link in 25e465cdb4e2d2b7f0c6db0d50ad1241b70df5ba

> You can customize the description with ``"cq_description_template"``, see `configuration file to pass options`_.

Author: Resolved in 9759d61e70f765a710954c3c50ef3ed07659644c

> Its value should be a template for Python's |string.Template|_. The template should contain ``$description`` and has
> access to all environment variables, e.g. ``$HOME``.
> This template functionality differs from the Polyspace checker.
> With the Polyspace checker, you can create custom descriptions using $-placeholders,
> which can represent environment variables or column titles from your TSV file.

Reviewer 2: ```suggestion
Its value should be a template for Python's |string.Template|_. The template has access to all environment variables,
e.g. ``$HOME``, and other variables that depend on the checker type:

Polyspace
  Any field of a Polyspace defect can be included by using the corresponding
  `column title <Exporting Polyspace Results>`_ in lowercase as the variable name.
  The default template is ``Polyspace: $check``
  
Other
  The template should contain ``$description``, which is the default.
```

-------------------

Use ``-C, --code-quality`` to let the plugin generate `a Code Quality report`_ for GitLab CI. All counted

> Sphinx, Doxygen, XMLRunner and Polyspace warnings/errors/failures will be included. Other checker types are not supported by this feature. The report is

Reviewer: Lets put some hope about other checkers.

```suggestion
Sphinx, Doxygen, XMLRunner and Polyspace warnings/errors/failures will be included. Other checker types are not yet supported by this feature. The report is
```
